---
title: "Session: Differential expression analysis with scRNA-seq data"
author: "Corinne Haines"
date: "2023-05-05"
output: html_document
---

# Session Description

This session will provide an introduction to differential expression analysis of
single-cell RNA-seq data.

# Session Learning Objectives

After completion, you should be able to perform differential expression analysis
using several Seurat methods and a pseudobulk method. Additionally, you should be
able to visualize the output of your differential expression analysis. Finally, 
you should have a general understanding of cell type annotation methods.

# Outline

* Load in data
    * Visualize data as a UMAP separated by treatment
* Differential Gene Expression
    * Between cell types
        * Change from clusters to cell type identities
        * FindAllMarkers()
        * Save markers as a .csv file
        * Visualize markers as a heatmap
        * Visualize markers as a dotplot
        * Visualize markers as a ridgeplot
    * *Activity (9:30-10ish)*
    * Within cell types between conditions
        * Find genes differentially expressed in DC cells between ctrl and stim
        * Explore the results
        * Save DEGs as a .csv file
        * Visualize data as a volcano plot
        * Visualize select genes as a FeaturePlot
        * Visualize select genes as a VlnPlot
    * Perform GSEA analysis on a list of DEGs
        * Save GSEA results as a .csv file
        * Visualize data as a bubble plot
    * *Activity (10:30-11ish)*
* Pseudobulked Differential Gene Expression
    * Why pseudobulk?
    * Prepare data for pseudobulk analysis
    * Perform DESeq2 on pseudobulk data
        * Save DEGs as a .csv file
        * Visualize data as a volcano plot
    * *Activity (11:15-11:30ish)*
* Brief intro to cell type annotation (description)
    * Identify conserved markers for cell type assignment 
    * Use singleR
    * Annotation with other scRNA-seq datasets

* Capstone projects:
    * Perform DEG analysis and visualization on clusters 
    * Perform pseudobulk analysis on clusters
    * Compare pseudobulk results to Seurat results

# Data

PBMC data from Lupus patients unstimulated or stimulated with interferon B (ifnb) (Kang et al 2017)

* Libraries were prepared using 10X Genomics v2 chemistry
* Samples were sequenced on the Illumina NextSeq 500
* PBMC samples from eight distinct lupus patients were separated into two aliquots each to receive different treatments:
    * 1 aliquot of PBMCs was activated by 100 U/mL of recombinant IFN-Î² for 6 hours
    * 1 aliquot was left untreated
* PBMCs that received the same treatment were then pooled together for processing as one 10X Genomics sample, then computationally demultiplexed:
    * After 6 hours, the eight samples for each condition were pooled together in two final pools (stimulated cells and control cells)
    * 12,138 and 12,167 cells were identified (after removing doublets) for control and stimulated pooled samples, respectively
    * Samples were demultiplexed using the tool Demuxlet
* Demo will be performed on full dataset
* Activities will be performed on subsetted dataset

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE
)

library(Seurat)
library(SeuratData)
library(ggplot2)
library(tidyverse)
library(stringr)
library(harmony)
library(patchwork)
library(msigdbr)
library(fgsea)
library(RColorBrewer)
library(ggrepel)
library(DESeq2)
library(pheatmap)

set.seed(888)


# not sure if we will need to do this or not:anchor.features
# options(future.globals.maxSize = 8000 * 1024^24)
```

```{r prepare_data, include = FALSE }
# prepare data using integration and SCTransformation

# look at which datasets are available
# AvailableData()

# install dataset of interest
# load dataset
data("ifnb")
# split data based on stim status
ifnb.list <- SplitObject(ifnb,
                    split.by = "stim")
# SCTransform (normalize, scale, find variable features) with glmGamPoi which
# speeds up the learning procedure
ifnb.list <- lapply(X = ifnb.list,
                    FUN = SCTransform,
                    method = "glmGamPoi")
# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = ifnb.list,
                    nfeatures = 3000)
# prepare object list normalized with sctransform for integration
ifnb.list <- PrepSCTIntegration(object.list = ifnb.list,
                    anchor.features = features)
# perform linear dimensional reduction
ifnb.list <- lapply(X = ifnb.list,
                    FUN = RunPCA,
                    features = features)
# identify "anchor" genes to integrate datasets
immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list,
                    normalization.method = "SCT",
                    anchor.features = features,
                    dims = 1:30,
                    reduction = "rpca",
                    k.anchor = 20)
# perform integration
immune.combined.sct <- IntegrateData(anchorset = immune.anchors,
                    normalization.method = "SCT",
                    dims = 1:30)
# perform linear dimensional reduction
immune.combined.sct <- RunPCA(immune.combined.sct,
                        verbose = FALSE)
# perform non-linear dimensional reduction  to learn the underlying manifold of
# the data in order to place similar cells together in low-dimensional space
immune.combined.sct <- RunUMAP(immune.combined.sct,
                        reduction = "pca",
                        dims = 1:30)
# we first construct a KNN graph based on the euclidean distance in PCA space,
# and refine the edge weights between any two cells based on the shared overlap
# in their local neighborhoods
immune.combined.sct <- FindNeighbors(immune.combined.sct,
                        dims = 1:30,
                        verbose = FALSE)
# apply modularity optimization techniques such as the Louvain algorithm to
# iteratively group cells together,
immune.combined.sct <- FindClusters(immune.combined.sct, verbose = FALSE)
data <- immune.combined.sct 
# save integrated and clustered dataset
saveRDS(data, "ifnb.rds")

# get sampleID information from 
# https://hbctraining.github.io/scRNA-seq_online/lessons/pseudobulk_DESeq2_scrnaseq.html
library(SingleCellExperiment)
sce_meta <- as.data.frame(colData(sce))
sce_meta$barcode <- gsub("-", ".", rownames(sce_meta))
sce_meta <- sce_meta[order(sce_meta$barcode),]
sce_meta <- sce_meta[,c("patient_id", "sample_id", "barcode")]
ifnb_meta <- as.data.frame(data@meta.data)
ifnb_meta$barcode <- rownames(ifnb_meta)
ifnb_meta <- ifnb_meta[order(ifnb_meta$barcode),]
merged <- merge(ifnb_meta, sce_meta)
# subset data for only those in both datasets
data <- subset(data, cells = merged$barcode)
data@meta.data <- data@meta.data[order(rownames(data@meta.data)),]
data@meta.data$sample_ID <- merged$sample_id
data@meta.data$patient_ID <- merged$patient_id
saveRDS(data, "data.rds")

# subset T-cells
Idents(data) <- "seurat_clusters"
lymphoid <- subset(data, idents = c("0","1","5","7","8","9","11","13","14","17","19","20"))
Idents(lymphoid) <- "seurat_annotations"
lymphoid <- subset(lymphoid, idents=c("CD4 Naive T", "CD4 Memory T",
                        "B", "T activated", "NK", "CD8 T", "B Activated"))
saveRDS(lymphoid, "ifnb_lymphoid.rds")

# subset myeloid cells
Idents(data) <- "seurat_clusters"
myeloid <- subset(data, idents = c("3", "2", "4", "6", "10", "12", "15", "18"))
Idents(myeloid) <- "seurat_annotations"
myeloid <- subset(myeloid, idents = c("CD14 Mono", "CD16 Mono", "DC", "pDC"))
saveRDS(myeloid, "ifnb_myeloid.rds")
```

# Load in data
```{r load_data}
# read in seurat objected
data <- readRDS("data.rds")

# set default assay
DefaultAssay(data) <- "RNA"
```

## Visualize data as a UMAP
```{r umap}
# set identity for the dataset
Idents(data) <- "seurat_annotations"
# plot UMAP separated by treatment
plot <- DimPlot(data, reduction = "umap", split.by = "stim")
```

# Differential Gene Expression

## Between cell types

### Find markers that define cell types
```{r}
# set identity to cell types
Idents(data) <- "seurat_annotations"
# find markers for each cell type
markers <- FindAllMarkers(data,
                min.pct = 0.25)
# reduce data
top_markers <- as.data.frame(markers %>% 
                group_by(cluster) %>%
                slice_max(avg_log2FC, n=5))
# save markers as a dataframe
write.csv(top_markers, "top5_markers.csv")
```

### Visualize markers as a heatmap
```{r}
# get list of top marker genes
genes <- unique(top_markers$gene)
# change to transformed assay
DefaultAssay(data) <- "SCT"
# make heatmap
heatmap <- DoHeatmap(data,
                features = genes,
                group.by = "seurat_annotations",
                size = 3)
```

### Visualize markers as a dotplot
```{r}
# reduce data
top_markers <- top_markers %>%
                    group_by(cluster) %>%
                    slice_max(avg_log2FC, n=3)
# get list of top marker genes
genes <- unique(top_markers$gene)
# make dot plot
dotplot <- DotPlot(data,
                features = genes) +
                RotatedAxis()
```

### Visualize markers as a ridgeplot
```{r}
# select a few genes
genes <- c("CCL2", "CCL5","HLA-DPB1", "GNLY")
# make ridgeplot
ridgeplot <- RidgePlot(data,
                features = genes,
                slot = "counts",
                log = TRUE,
                ncol = 2)
```

## Activity

### Read in T-cell data, set assay to the RNA slot, and set idents to clusters:
```{r}
```

### Visualize clusters as a UMAP split by treatment
```{r}
```

### Find markers that define clusters and visualize top 2 markers using a dotplot
```{r}
```








## Within cell types between conditions

### Find genes differentially expressed in DC cells between ctrl and stim
```{r}
Idents(data) <- "seurat_annotations"
# Find DEGs for DC cells
DC_degs <- FindMarkers(data,
                    ident.1 = "STIM",
                    ident.2 = "CTRL",
                    group.by = "stim",
                    subset.ident = "DC",
                    min.pct = 0.1,
                    logfc.threshold = 0.25)
```

### Explore the results
```{r}
# look at first 20 differentially expressed genes
head(DC_degs, n=20)

# how many significantly different genes are there?
nrow(DC_degs[DC_degs$p_val_adj <= 0.05,])
```

The results data frame has the following columns :

* p_val : p_val (unadjusted)
* avg_log2FC : log fold-change of the average expression between the two groups. 
** Note : positive values indicate that the feature is more highly expressed in the first group.
* pct.1 : percentage of cells where the feature is detected in the first group
* pct.2 : percentage of cells where the feature is detected in the second group
* p_val_adj : adjusted p-value, based on Bonferroni correction using all features in the dataset.

### Save DEGs as a .csv file
```{r}
write.csv(DC_degs, file = "DC_stim_vs_ctrl_degs.csv")
```

### Visualize data as a volcano plot
```{r}
# log transform the padj values for visualization
DC_degs$log.padj <- -log10(DC_degs$p_val_adj)
# pull out labels for the top 20 most differentially expressed genes
labels <- DC_degs %>%
                arrange(desc(abs(avg_log2FC))) %>%
                slice_head(n=20)
# make volcano plot
volcano <- ggplot(DC_degs, aes(x=avg_log2FC, y=log.padj)) +
                geom_point() +
                geom_label_repel(data = labels,
                    aes(label = rownames(labels)),
                    max.overlaps = 20) +
                ggtitle("Dendritic Cell Differentially Expressed Genes") +
                ylab("-log(padj)") +
                xlab("log2FC") +
                theme(plot.title = element_text(hjust = 0.5),
                    text = element_text(size = 12))
```

### Visualize select genes as a FeaturePlot
```{r}
# make feature plot
feature <- FeaturePlot(data,
                features = c("IL8", "ISG15", "CXCL10"),
                split.by = "stim",
                order = TRUE)
```

### Visualize select genes as a VlnPlot
```{r}
# make violin plot
violin <- VlnPlot(data,
                features = c("IL8", "ISG15", "CXCL10"),
                split.by = "stim",
                group.by = "seurat_annotations",
                ncol = 1,
                pt.size = 0)
```

## Perform GSEA analysis on the DC differentially expressed genes
```{r}
# look at genesets available
msigdbr_collections()
# select dataset
GO.MF <- msigdbr(species = "Homo sapiens",
                category = "C5",
                subcategory = "GO:MF") %>%
                split(x = .$gene_symbol,
                f = .$gs_name)
# arrange DEGs for GSEA
DC_degs$gene <- row.names(DC_degs)
deg <- DC_degs %>%
            arrange(desc(avg_log2FC)) %>%
            dplyr::select(gene, avg_log2FC)
ranks <- deframe(deg)
# perform GSEA
fgseaRes <- fgseaMultilevel(GO.MF,
                ranks,
                minSize = 15,
                maxSize = 500,
                nPerm = 1000) %>%
            arrange(padj)
# take a look at results
head(fgseaRes)
# save results as a .csv file
write.csv(fgseaRes, "DC_stim_vs_ctrl_GSEA.csv")
```

### Visualize data as a bubble plot
```{r}
# get only the top 10 pathways
fgseaRes <- fgseaRes[1:10,]
# make bubble plot
bubble <- ggplot(fgseaRes, 
                aes(x = NES, 
                    y=pathway, 
                    size = size,
                    color = padj)) +
                geom_point() +
                scale_color_gradient2(high = "#f517d3",
                    mid = "#ffffff",
                    low = "#0513d8",
                    midpoint = 0.05)
```

## Activities

### Find genes differentially expressed in activated T cells between ctrl and stim
```{r}
```
### Explore the results and save as a .csv file
```{r}
```
### Visualize top 3 most significantly different genes using FeaturePlot
```{r}
```
### Visualize top 3 most significantly different genes using VlnPlot
```{r}
```
## Perform GSEA analysis on the activated T cell DEGs using the Hallmark geneset
```{r}
```
### Visualize GSEA results as a bubble plot
```{r}
```







# Pseudobulked Differential Gene Expression
## Why pseudobulk?
The DE analysis in Seurat treats each cell as an independent sample often 
resulting in inflated p-values as single cells isolated from the same biological
sample are not *truly* independent of one another. Thus, the methods used to
perform DE analysis in Seurat identify genes that vary across individual cells 
not necessarily across a population. To infer DE genes across a population between
conditions we need to pseudobulk our cells.

## Prepare data for pseudobulk analysis
First, generate DEG tables from pseudobulk clusters, treating all cells
from each cluster from each sample as independent groups of cells/data.
Then, graph each of those DEG datasets using a volcano plot.
```{r}
# Create pseudobulk tables

# create column in metadata the reflects stim, cellType, and patient ID
data@meta.data$stim_cell_ID <- gsub(" ", "_", paste0(data@meta.data$stim, "_",
                                data@meta.data$seurat_annotations, "_",
                                data@meta.data$patient_ID))
# create column in metadata the reflects stim and cellType
data@meta.data$stim_cell <- gsub(" ", "_", paste0(data@meta.data$stim, "_",
                                data@meta.data$seurat_annotations))
# change Idents
Idents(data) <- "stim_cell_ID"
# get count data
pb <- data.frame(gene = rownames(data@assays$RNA@counts))
# set rownames as gene names
rownames(pb) <- pb$gene
# create pseudobulked data for ctrl and stim cell types for each patient
for (sample in levels(as_factor(data@meta.data$stim_cell_ID))){
    # subset data to include a single ctrl/stim cell type for a single patient
    sub <- subset(data, subset = stim_cell_ID == sample)
    # include data only if there are more than 10 cells for that sample
    if (nrow(sub@meta.data) > 9){
        # sum up the counts for each gene and save it to the pseudobulk dataframe
        pb[sample] <- rowSums(sub@assays$RNA@counts)
    } else {
        print(paste("NOTE: sample", sample, "has less than 10 cells. 
        This sample will be skipped."))
    }
}
# remove gene column (since rownames are genes)
pb <- pb[, -1]
# save pseudobulk data as a dataframe
write.csv(pb, "pseudobulk.csv")

# Create a separate column (meta) data table as required for DESeq2 input
# make a dataframe with a sample column that has 
# the columns from the pseudobulk dataframe
colData <- data.frame(sample = colnames(pb))
# pull out just the patient ID and save that to its own column
colData$patient_ID <- str_sub(colData$sample, 
                start=nchar(colData$sample)-3,
                end = nchar(colData$sample)) %>%
                gsub("_", "", .)
# create a column that is the stim_cell type
for (patient in levels(as_factor(colData$patient_ID))){
    for (cell in levels(as_factor(data$seurat_annotations))){
            modCell <- gsub(" ", "_", cell)
        for (stim in levels(as_factor(data$stim))){
            sample <- paste0(stim, "_", modCell, "_", patient)
            colData[colData$sample == sample, "stim_cell"] <- paste0(stim, "_", modCell)
        }
    }
}
# create a column that is just the cell type
colData$cellType <- str_sub(colData$stim_cell, 
                start=6,
                end = nchar(colData$stim_cell))
# set rownames as the sample names
rownames(colData) <- colData$sample
# save the colData as a .csv file
write.csv(colData, "colData_for_DESeq2.csv")
```

## Perform DESeq2 between CTRL vs STIM within each cellType
```{r}
# set up DDS object using ~stim_cell as the design 
dds <- DESeq2::DESeqDataSetFromMatrix(
            countData = pb,
            colData = colData,
            design = ~ stim_cell)
# run DESeq2
dds <- DESeq2::DESeq(dds)
# save DDS object
saveRDS(dds, "dds.rds")

# make a list for the DEGs to go into
deseq2 <- list()
# pull out DEGs for each cell type
for (cell in unique(colData$cellType)){
    # make a variable for the ctrl cell type
    ctrl <- paste0("CTRL_", cell)
    # make a variable for the stim cell type
    stim <- paste0("STIM_", cell)
    # pull out the differentially expressed genes such that the stim cell type
    # is the numerator and the ctrl cell type in the denominator
    res <- as.data.frame(DESeq2::results(dds,
                                contrast = c("stim_cell", stim, ctrl)))
    # remove any NAs in the data
    res <- res[!is.na(res$padj),]
    # re-order the data based on padj
    res <- res[order(res$padj),]
    # store the results in the DEG list
    deseq2[[cell]]$results <- res
}
```


### Visualize data as volcano plots
```{r}
# make a list for the plots to go into
plots <- list()
for (cell in unique(colData$cellType)) {
    # pull out the names of the genes that are most significantly DE
    labs <- head(deseq2[[cell]]$results, n = 15)
    # make a violin plot for the DEGs for each cell type
    plots[[cell]]$volcano <-
        ggplot(deseq2[[cell]]$results, aes(x = log2FoldChange, y=-log10(padj))) +
            geom_point() +
            geom_label_repel(
                data = labs,
                aes(label = rownames(labs)),
                max.overlaps = 20) +
            ggtitle(paste(cell, "DEGs")) +
            ylab("- log ( padj )") +
            xlab("log2FC (STIM / CTRL)") +
            theme(plot.title = element_text(hjust = 0.5),
                    text = element_text(size = 12))
}
```

## Activities

### Create pseudobulk dataset
```{r}
```

### Run DESeq2
```{r}
```
### Pull out the T activated DEGs and plot a Volcano plot
```{r}
```







# Brief intro to cell type annotation (description)
## Identify conserved markers for cell type assignment
Seurat's FindConservedMarkers() function will identify genes which are
differentially expressed in each identity group (in this case, clusters) in
comparison to all other identity groups AND are similarly expressed (conserved)
between the groups you are comparing (in this case, ctrl and stim). 

We can use Seurat's FindAllMarkers() function to identify genes which are 
differentially expressed in each identity group (in this case, clusters) in 
comparison to all other identity groups. This method may identify genes which are
markers for multiple closely-related groups (i.e. not specific to one group).

You can then manually define the cluster cell type identities.
```{r manual, eval = FALSE}
# set identity of object to be clusters
Idents(data) = "seurat_clusters"

# loop through each cluster number with FindConservativeMarkers & save data
for (i in 0:(length(levels(data$seurat_clusters))-1)){
    # find conserved markers for each cluster irrespective of stimulation
    conserved=FindConservedMarkers(data, ident.1 = i, grouping.var = "stim")
    # create file name
    filename=paste0("./conservedMarkers/Cluster", i, "_conservedMarkers.csv")
    # save markers as a .csv file
    write.csv(conserved, file = filename)
}

# run FindAllMarkers & save data for each cluster as an indiviudal file
markers <- FindAllMarkers(data)
for (i in 0:(length(levels(data$seurat_clusters))-1)){
    df <- markers[markers$cluster == i,]
    filename=paste0("./markers/Cluster", i, "_markers.csv")
    write.csv(df, file = filename)
}

# add cell type ids to metadata
Idents(data) <- "seurat_clusters"
data <- RenameIdents(data, `0` = "CD14 Mono", `1` = "CD4 Naive T", 
                    `2` = "CD4 Memory T", `3` = "CD16 Mono", `4` = "B", 
                    `5` = "CD8 T", `6` = "NK", `7` = "T activated", `8` = "DC", 
                    `9` = "B Activated", `10` = "Mk", `11` = "pDC", 
                    `12` = "Eryth", `13` = "Mono/Mk Doublets", `14` = "HSPC")
# remember to save your data!
```

## Use singleR
We can also use the SingleR package and cell type references to annotate our data.

https://bioconductor.org/packages/release/bioc/html/SingleR.html

```{r singleR, eval = FALSE}
library(SingleR)
# singleR allows you to use celldex cell type references (can use others)
library(celldex)
# get count data for your scRNA-seq data
counts <- GetAssayData(data)
# load the celldex data you want to use
ImmGenData <- ImmGenData()
# run singleR with your count data and the cellDex data
pred <- SingleR(test = counts, ref = ImmGenData, 
    labels = ImmGenData$label.fine, assay.type.test=1)
# add the predicted cell types to your metadata
data[["ImmGen.labels"]] <- pred$labels
```

## Annotation with other scRNA-seq datasets
We can also download other people's scRNA-seq data from public repositories and
use their metadata to annotate our own. This method makes use of gene set scoring.
This method can be used for cell-typing data or you could also use it for any 
other metadata column (sex, treatment, age, etc.)

```{r TransferAnchors, eval = FALSE}
# load reference 
coolReference.seurat <- load(coolReference.RData)
ref <- coolReference.seurat
# note: reference and query dataset need to be analyzed with the same 
# normalization method, so you may have to extract raw data and then re-normalize
# set identity to your goal idents
Idents(ref) <- ref$cellIDs
# visualize
DimPlot(ref, reduction = "umap")
# align assays
DefaultAssay(ref) <- "RNA"
DefaultAssay(myData) <- "RNA"
# find anchors (use same normalization method and dims for both datasets)
anchors <- FindTransferAnchors(reference = ref,
                query = myData,
                normalization.method = "SCT",
                dims = 1:30)
# transfer data
predictions <- TransferData(anchorset = anchors,
                    refdata = ref$cellIDs,
                    dims = 1:30)
# add predictions to metadata
myData <- AddMetaData(myData, metadata = predictions)
# move data to informative column
myData$ref_labels <- myData$predicted.id
# change idents to predicted id
Idents(myData) <- "ref_labels"
```


# Capstone projects:
## Perform DEG analysis and visualization on clusters
```{r}
```

## Perform pseudobulk and DEG analysis on clusters
```{r}
```

## How do the results of the pseudobulk analysis compare to the Seurat analysis?
```{r}
```








## Lymphoid pseudobulk and DESeq2
```{r, include = FALSE}
# Create pseudobulk tables 

# create column in metadata the reflects stim, cellType, and patient ID
data@meta.data$stim_cell_ID <- gsub(" ", "_", paste0(data@meta.data$stim, "_",
                                data@meta.data$seurat_annotations, "_",
                                data@meta.data$patient_ID))
# create column in metadata the reflects stim and cellType
data@meta.data$stim_cell <- gsub(" ", "_", paste0(data@meta.data$stim, "_",
                                data@meta.data$seurat_annotations))
# change Idents
Idents(data) <- "stim_cell_ID"
# get count data
pb <- data.frame(gene = rownames(data@assays$RNA@counts))
# set rownames as gene names
rownames(pb) <- pb$gene
# create pseudobulked data for ctrl and stim cell types for each patient
for (sample in levels(as_factor(data@meta.data$stim_cell_ID))){
    # subset data to include a single ctrl/stim cell type for a single patient
    sub <- subset(data, subset = stim_cell_ID == sample)
    # include data only if there are more than 10 cells for that sample
    if (nrow(sub@meta.data) > 9){
        # sum up the counts for each gene and save it to the pseudobulk dataframe
        pb[sample] <- rowSums(sub@assays$RNA@counts)
    } else {
        print(paste("NOTE: sample", sample, "has less than 10 cells. 
        This sample will be skipped."))
    }
}
# remove gene column (since rownames are genes)
pb <- pb[, -1]
# save pseudobulk data as a dataframe
write.csv(pb, "lymphoid_pseudobulk.csv")

# Create a separate column (meta) data table as required for DESeq2 input
# make a dataframe with a sample column that has 
# the columns from the pseudobulk dataframe
colData <- data.frame(sample = colnames(pb))
# pull out just the patient ID and save that to its own column
colData$patient_ID <- str_sub(colData$sample, 
                start=nchar(colData$sample)-3,
                end = nchar(colData$sample)) %>%
                gsub("_", "", .)
# create a column that is the stim_cell type
for (patient in levels(as_factor(colData$patient_ID))){
    for (cell in levels(as_factor(data$seurat_annotations))){
            modCell <- gsub(" ", "_", cell)
        for (stim in levels(as_factor(data$stim))){
            sample <- paste0(stim, "_", modCell, "_", patient)
            colData[colData$sample == sample, "stim_cell"] <- paste0(stim, "_", modCell)
        }
    }
}
# create a column that is just the cell type
colData$cellType <- str_sub(colData$stim_cell, 
                start=6,
                end = nchar(colData$stim_cell))
# set rownames as the sample names
rownames(colData) <- colData$sample
# save the colData as a .csv file
write.csv(colData, "lymphoid_colData_for_DESeq2.csv")

# set up DDS object using ~stim_cell as the design 
dds <- DESeq2::DESeqDataSetFromMatrix(
            countData = pb,
            colData = colData,
            design = ~ stim_cell)
# run DESeq2
dds <- DESeq2::DESeq(dds)
# save DDS object
saveRDS(dds, "lymphoid_dds.rds")
```