---
title: "Coder Upgrade 2023: Multiomics analysis"
author: "Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      cache = TRUE,
                      cache.lazy = FALSE)
```

Load libraries
```{r libraries, cache=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(rrrSingleCellUtils)
library(tidyverse)
library(Seurat)
library(Signac)
library(nichenetr)
#system("ml HDF5")
```

--------------------------------------------------------------------------------

# Section 1: Overview

## What is multiomics single-cell data?

Any technology that produces multiple type of data from single cells
Each cell is profiled for multiple types of data

- RNA / ATAC
- RNA / ChIP
- RNA / Protein
- RNA / DNA
- RNA / DNA methylation / Chrom. accessibility

## Benefits of multiomics

- Direct association within individual cells between different types of data
- Avoids shakey correlations made using bulk methods on heterogeneous populations
- Can generate insights not possible with other methods

## Challenges of multiomics

- Generally, the data quality will be much worse for each individual cell compared to bulk methods
- Sample/library preparation is more complex
    - Especially for homebrew methods
- $$$
- Analysis is going to be much more complex
    - Need to analyze multiple single-cell data types
    - Need proper statistical methods to integrate the data
    - Methods are generally less mature than more commonly used methods

## 10x multiomics kit

- Generates both RNA and ATAC data from single nuclei
- Relies on cleanly isolated and fixed nuclei
    - Get only RNAs from the nucleus
    - Nuclei prep can be quite challenging
    - Likely more prone to cell-cell cross contamination

## 10x multiomics data structure

RNA and ATAC have different sequencing requirements

- RNA
    - R1: 28bp <- has UMI and nucleus barcode
    - R2: 90bp <- has RNA sequence
    - I1/I2: 10bp <- have sample barcodes
- ATAC
    - R1: 50bp <- has genomic sequence
    - R2: 24bp <- has nucleus barcode
    - R3: 49bp <- has genomic sequence
    - I1: 8bp <- has sample barcode
- Recommended 20,000 read pairs / nucleus for GEX
    - In 10x demo dataset, 122,335 reads per nucleus
- Recommended 25,000 read pairs / nucleus for ATAC
    - In 10x demo datasets, 95,897 reads per nucleus
- Aim higher if possible
    - For 2000 cells, 100k + 100k reads per sample
        - 400M reads per sample
        - ~4 samples on a NovaSeq S1

## Cellranger-arc

Depending on what data type you recieve you may need to do different things

- BCLs
    - Run cellranger-arc mkfastq
    - If both GEX and ATAC run on single lane, need to specify read lengths for each
    - If run seperately on different lanes, can just run mkfastq on each lane
- Fastqs
    - Run cellranger-arc count
    - Simple way is to do one run per sample
        - Input is csv with sample info
    - Output is a folder with a bunch of files/folders
        - https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/using/count
        - web_summary.html              - html report
        - filtered_feature_bc_matrix/   - contains the GEX data
        - filtered_feature_bc_matrix.h5 - h5 file with both GEX and ATAC data
        - atac_peaks.bed                - ATAC peaks
        - atac_fragments.tsv.gz         - ATAC fragments
        - cloupe.cloupe                 - Loupe browser file
        - *.bam                         - BAM files

--------------------------------------------------------------------------------

## Activity (9:30 - 10ish)

Look at output of cellranger-arc count
Create a folder of data to look at and put it in gdworkshop/data/...

--------------------------------------------------------------------------------

## Using Seurat and Signac for 10x multiomics data individually

You essentially get two independant datasets back (GEX and ATAC) that can be analyzed independently.

The GEX data can be analyzed like normal

- One exception - if you read in your data using `Read10X_h5()`, it returns a list of two objects:
    - Gene Expression
    - Peaks
- Need to pull the GEX data out of the list and use that for analysis

I'm not going to walk through the GEX analysis as that's been covered in previous sessions
```{r}
data_path <- "/home/gdworkshop/lab/session_data/multiomics/"
h5_data <-
    Read10X_h5(paste0(data_path,
                      "e18_mouse_brain_fresh_5k_filtered_feature_bc_matrix.h5"))

# mt_pattern <- "^mt-" because this is mouse data
seurat_obj <-
    CreateSeuratObject(counts = h5_data$`Gene Expression`,
                       assay = "RNA",
                       project = "10x_multiomics",
                       min_cells = 5,
                       min_features = 800) %>%
    PercentageFeatureSet(pattern = "^mt-",
                         col.name = "percent_mt_rna",
                         assay = "RNA")
```


## Add the ATAC data to the Seurat object
Need to combine the datasets prior to filtering
```{r}
frag_file <-
    paste0(data_path,
           "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz")

seurat_obj[["ATAC"]] <-
    CreateChromatinAssay(counts = h5_data$Peaks,
                         sep = c(":", "-"),
                         fragments = frag_file,
                         min.cells = 5)
```

# Filter out likely junk cells
```{r}
cutoffs <-
    tribble(~feature, ~min_val, ~max_val,
            "nCount_RNA", 2000, 25000,
            "nFeature_RNA", 1000, 7000,
            "percent_mt_rna", 0, 20,
            "nCount_ATAC", 2000, 60000,
            "nFeature_ATAC", 1000, 25000)

feature_hist(seurat_obj,
             features = c("nCount_RNA",
                          "nFeature_RNA",
                          "percent_mt_rna",
                          "percent_mt_atac",
                          "nCount_ATAC",
                          "nFeature_ATAC"),
             cutoff_table = cutoffs)

length(Cells(seurat_obj))
seurat_obj <-
    subset(seurat_obj,
           subset = nCount_RNA >= cutoffs$min_val[1] &
                    nCount_RNA <= cutoffs$max_val[1] &
                    nFeature_RNA >= cutoffs$min_val[2] &
                    nFeature_RNA <= cutoffs$max_val[2] &
                    percent_mt_rna <= cutoffs$max_val[3]) %>%
    process_seurat(resolution = 0.2)
length(Cells(seurat_obj))

# Not the best way to characterize brain cell types, but it's good enough for now
imm_cells <- celldex::ImmGenData()
mouse_rna <- celldex::MouseRNAseqData()

cell_assign <-
    SingleR::SingleR(as.SingleCellExperiment(seurat_obj),
                     ref = list(imm_cells,
                                mouse_rna),
                     labels = list(imm_cells$label.fine,
                                   mouse_rna$label.fine))

seurat_obj$cell_type <-
    cell_assign$labels %>%
    str_replace("DC", "Dendritic cells")
seurat_obj$cell_score <-
    cell_assign$scores %>%
    apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

DimPlot(seurat_obj, group.by = "cell_type", label = TRUE)

# Save this so I don't have to re-run everything
qs::qsave(seurat_obj, paste0(data_path,
                             "mouse_brain_5k_sobj_1.qs"))
```

# Let's focus on the ATAC data

Add in gene annotations
```{r}
DefaultAssay(seurat_obj) <- "ATAC"

annotations <-
    GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79::EnsDb.Mmusculus.v79)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "mm10"

# add the gene information to the object
Annotation(seurat_obj[["ATAC"]]) <- annotations
```

Add in nucleosome signal. This is a the ratio of mononucleosome fragments to short fragments.
```{r}
seurat_obj <- NucleosomeSignal(seurat_obj)
seurat_obj$nucleosome_group <-
    ifelse(seurat_obj$nucleosome_signal > 4,
           paste0("NS > ", 4),
           paste0("NS < ", 4))
```

Calculate the signal enrichment at transcription start sites
```{r}
seurat_obj <- TSSEnrichment(seurat_obj, fast = FALSE)
seurat_obj$high_tss <-
    ifelse(seurat_obj$TSS.enrichment > 2,
           "High",
           "Low")
```

Calculate how many of the reads are in types of regions
```{r}
total_frag_df <-
    CountFragments(paste0(data_path,
                          "e18_mouse_brain_fresh_5k_atac_fragments.tsv.gz"))

total_frag_df <-
    total_frag_df %>%
    dplyr::filter(CB %in% colnames(seurat_obj)) %>%
    dplyr::arrange(match(CB, colnames(seurat_obj)))

# Populate seurat_obj with metadata

seurat_obj$total_frag <- total_frag_df$reads_count
seurat_obj$mononucleosomal <- total_frag_df$mononucleosomal
seurat_obj$nucleosome_free <- total_frag_df$nucleosome_free
```

Calculate the fraction of reads in peaks
```{r}
seurat_obj <-
    FRiP(seurat_obj,
         assay = "ATAC",
         total.fragments = "total_frag",
         col.name = "FRiP",
         verbose = FALSE)
```

Add in metadata output by cellranger-arc count
```{r}
metadata <-
    read_csv(paste0(data_path, "/e18_mouse_brain_fresh_5k_per_barcode_metrics.csv"),
             show_col_types = FALSE) %>%
    mutate(atac_pct_mito = atac_mitochondrial_reads / atac_raw_reads * 100,
           atac_pct_dups = atac_dup_reads / atac_raw_reads * 100) %>%
    filter(barcode %in% colnames(seurat_obj)) %>%
    select(-atac_barcode,
           -is_cell,
           -excluded_reason,
           -gex_barcode) %>%
    arrange(match(barcode, colnames(seurat_obj))) %>%
    column_to_rownames("barcode")

seurat_obj <- AddMetaData(seurat_obj, metadata)

# Save it so I don't have to re-run everything
qs::qsave(seurat_obj, paste0(data_path,
                             "mouse_brain_5k_sobj_2.qs"))
```

Lets plot our new metadata
```{r}
plot_factors <-
    c("TSS.enrichment",
      "FRiP",
      "nucleosome_signal",
      "atac_pct_mito",
      "nCount_ATAC",
      "atac_pct_dups")
plot_name <- feature_hist(seurat_obj, plot_factors)
ggsave("Sessions/08/atac_feat_hist.png",
       plot_name,
       width = 6,
       height = 10)
```

One issue:

If you have multiple samples called with cellranger-arc, their peaks are different, so you can't really compare them directly.

You can either re-call the peaks: https://stuartlab.org/signac/articles/peak_calling.html

Or redefine their boundaries: https://stuartlab.org/signac/articles/merging.html

https://github.com/kidcancerlab/rrrSingleCellUtils/blob/master/R/multiomics.R#L18

```{r}
# Documentation is still a work in progress
library(rrrSingleCellUtils)
peak_beds <- paste0(data_path,
                    c("sample_1/atac_peaks.bed",
                      "sample_2/atac_peaks.bed",
                      "sample_3/atac_peaks.bed"))
frag_paths <- paste0(data_path,
                     c("sample_1/atac_fragments.tsv.gz",
                       "sample_2/atac_fragments.tsv.gz",
                       "sample_3/atac_fragments.tsv.gz"))
cell_ids <- c("sample_1",
              "sample_2",
              "sample_3")

# Uses {parallel} to run faster
merged_atac_data <-
    merge_atac(peak_beds = peak_beds,
               frag_paths = frag_paths,
               cell_ids = cell_ids,
               n_regions_simul = 100000,
               threads = 3) %>%
    subset(nCount_ATAC > 100)
```

https://stuartlab.org/signac/articles/merging.html

## Activity (10:30 - 11ish)

Work with Seurat and Signac

## Joint analysis of multimodal data

> Show ppt of plots

- Multimodal UMAP
    - Better separation of cell types due to increased dimensionality of data
- Correlation of ATAC peaks with gene expression
    - Identify active chromatin that might influence gene expression
    - Motif analysis of peak sequence can potentially identify binding factors
- Compare heterogeneity of gene expression vs. chromatin accessibility

## Activity (11:30 - 12ish)

Talk to Amy, Meren and Jason Navarro?

## Session challenge





# Resources
Methods and applications for single-cell and spatial multi-omics: https://www.nature.com/articles/s41576-023-00580-2

Into the multiverse: advances in single-cell multiomic profiling: https://www.cell.com/trends/genetics/fulltext/S0168-9525(22)00077-4

https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression

https://stuartlab.org/signac/

Merging ATAC datasets: https://stuartlab.org/signac/articles/merging.html